include/pybind11/functional.h

```diff
@@ -50,9 +50,9 @@ struct type_caster<std::function<Return(Args...)>> {
         }
 
         auto src = reinterpret_borrow<object>(src_);
-        value = [src](Args... args) -> Return {
+        value = [src](Args&&... args) -> Return {
             gil_scoped_acquire acq;
-            object retval(src(std::move(args)...));
+            object retval(src(std::forward<Args>(args)...));
```

Comment:
Nice. Looks like this will fix the behavior for `T&` in general, not just with abstract types. For example, the following test fails prior to this change:

```c++
struct MovableObject {
    bool valid = true;

    MovableObject() = default;
    MovableObject(const MovableObject &) = default;
    MovableObject &operator=(const MovableObject &) = default;
    MovableObject(MovableObject &&o) : valid(o.valid) { o.valid = false; }
    MovableObject &operator=(MovableObject &&o) { valid = o.valid; o.valid = false; return *this; }
};

py::class_<MovableObject>(m, "MovableObject");

m.def("callback_with_movable", [](std::function<void(MovableObject &)> f) {
    auto x = MovableObject();
    f(x); // lvalue reference shouldn't move out object
    return x.valid; // must still return `true`
});
```

```python
assert callback_with_movable(lambda _: None) is True  # returns `False` without `std::forward`
```